wdl-analysis/src/types/v1.rs

```diff
@@ -531,6 +533,14 @@ pub struct ExprTypeEvaluator<'a, C> {
     /// If the count is non-zero, special evaluation behavior is enabled for
     /// string interpolation.
     placeholders: usize,
+
+    /// A cache of regexes that have been evaluated.
+    ///
+    /// This is used to avoid re-evaluating the same regex multiple times.
+    ///
+    /// The key is the regex string and the value is either an error or
+    /// `Ok(())`.
+    regex_cache: HashMap<String, Result<(), String>>,
```

Comment:
this doesn't actually save much compute time. Instead of caching `Ok(())`, we should cache the _compiled regex_ (IDK what that Rust type is, but it's something offered by the `regex` crate).

The expensive operation here is (potentially) compiling the regex (if it's sufficiently complicated, which is admittedly a pretty niche case). Once it's compiled, using the RE to search text is a relatively light operation. So we want to compile once to ensure it's valid, and then cache the compiled RE for re-use during `run` operations.

If someone reviewing this knows more than me and something I said here is wrong, please correct :)